pkg/datasource/opensergo/opensergo.go

```diff
@@ -0,0 +1,122 @@
+// Copyright 1999-2020 Alibaba Group Holding Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package opensergo
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/alibaba/sentinel-golang/ext/datasource"
+	"github.com/alibaba/sentinel-golang/logging"
+	"github.com/alibaba/sentinel-golang/util"
+	"github.com/opensergo/opensergo-go/pkg/client"
+	"github.com/opensergo/opensergo-go/pkg/configkind"
+	"github.com/opensergo/opensergo-go/pkg/transport/subscribe"
+	"github.com/pkg/errors"
+)
+
+type OpenSergoDataSource struct {
+	datasource.Base
+	isInitialized           util.AtomicBool
+	client                  client.OpenSergoClient
+	namespace               string
+	app                     string
+	opensergoRuleAggregator *OpensergoRuleAggregator
+}
+
+func NewOpenSergoDataSource(host string, port int, namespace string, app string) (*OpenSergoDataSource, error) {
+	if len(namespace) == 0 || len(app) == 0 {
+		return nil, errors.New(fmt.Sprintf("invalid parameters, namespace: %s, app: %s", namespace, app))
+	}
+	ds := &OpenSergoDataSource{
+		client:                  *client.NewOpenSergoClient(host, port),
+		namespace:               namespace,
+		app:                     app,
+		opensergoRuleAggregator: NewOpensergoRuleAggregator(),
+	}
+
+	ds.AddPropertyHandler(datasource.NewDefaultPropertyHandler(MixedPropertyJsonArrayParser, MixedPropertyUpdater))
```

Comment:
MixedPropertyJsonArrayParser is a must?  Can the parser be loaded on demand？

Reply:
> MixedPropertyJsonArrayParser is a must? Can the parser be loaded on demand？

yeah. It is a must one. By considing the mixed rule-types returned in func ReadSouce() for shared gRpc client mode.
`shared gRpc client mode` refer to https://github.com/opensergo/opensergo-java-sdk/issues/8

Through MixedPropertyJsonArrayParser() and MixedPropertyUpdater() in mixed_property_adaptor.go, we can handle both single rule-type and mixed rule-types.

If sentinel-go want to keep the same with other data-source, i iwill re-code it.

Reply:
re-code it, but still with MixedPropertyJsonArrayParser() and MixedPropertyUpdater()